File: leetcode/0133-clone-graph/solution.cpp
#include <unordered_map>
#include <vector>
using namespace std;

class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() { val = 0; }
    Node(int _val) { val = _val; }
    Node(int _val, vector<Node*> _neighbors) { val = _val; neighbors = _neighbors; }
};

class Solution {
public:
    unordered_map<Node*, Node*> cloned;

    Node* cloneGraph(Node* node) {
        if (!node) return nullptr;

        auto it = cloned.find(node);
        if (it != cloned.end()) return it->second;

        Node* copy = new Node(node->val);
        cloned[node] = copy;
        for (Node* neighbor : node->neighbors) {
            copy->neighbors.push_back(cloneGraph(neighbor));
        }

        return copy;
    }
};

File: leetcode/0133-clone-graph/solution.py
from typing import Optional


# Definition for a graph node.
class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if node is None:
            return None

        cloned = {}

        def dfs(orig: 'Node') -> 'Node':
            if orig in cloned:
                return cloned[orig]

            copy = Node(orig.val)
            cloned[orig] = copy
            for neighbor in orig.neighbors:
                copy.neighbors.append(dfs(neighbor))

            return copy

        return dfs(node)
